src/org/rascalmpl/library/analysis/m3/Core.rsc

```diff
@@ -34,54 +34,157 @@ import List;
+	rel[loc definition, loc comments] documentation  = {},	
+	rel[loc definition, Modifier modifier] modifiers = {}	
+) = m3(loc id);
              
 
+@synopsis{Extensible data-type to define language names and their versions}
+@description{
+Most ground truth facts about source code require analysis tooling that is specific to the language:
+* parsers
+* name analysis
+* type analysis
+
+However, there are language analysis methods that are _language agnostic_ such as counting lines of code.
+For this we have the `generic()` language name.
+}
 public data Language(str version = "")
   = generic()
   ;
 
-@synopsis{Create an empty m3 term with empty annotations}
-//TODO: Deprecated method, replace any calls to this method with default constructor
-public M3 emptyM3(loc id) = m3(id);
-
-@synopsis{Generic function to compose the annotations of a set of M3s.}
+@synopsis{Generic function to compose the facts of a set of M3s into a single model.}
+@description{
+We iterate over all the facts stored in every model, and use set union or list concatenation
+to collect the elements of all relations and lists. 
+}
+@pitfalls{
+* If the quality of the qualified names in the original models is lacking, than this is the moment that different
+declarations might be conflated with the same fully qualified name. All downstream analysis is broken then.
+* This function does not compose the extended facts for specific programming languages yet.
+* If extended M3 models use something other than sets, lists or relations, this composition function ignores them completely.
+* Composed models can be huge in memory. Make sure to allocate enough heap for the JVM. Real world programs of real world product
+can take gigabytes of memory, even when compressed and optimized as M3 models. 
+}
+@benefits{
+* Composition satisfies the requirements for many downstream analyses:
+   * Composition can be used to easily construct project-level models from file-level models, e.g. for open-source project analysis.
+   * Composition can be used to simulate (dynamic) linkage between projects, e.g. for whole-program analysis.
+   * Composition can be used to start simulating remote-procedure calls and shared memory, and other inter-programming language composition like _JNI_.
+* Transitive closure on composed models leads to effective (and fast) reachability analysis.
+* This function is rather memory-efficient by iterating over the already in-memory keyword parameter sets and lists, and
+splicing the unions into an incremental transiently growing set or list via the comprehension. This avoids a lot of copying
+and intermediate memory allocation which can be detrimental when doing large whole program analyses.
+}
 @memo
 M3 composeM3(loc id, set[M3] models) {
 	M3 comp = m3(id);
+  map[M3, map[str,value]] fields = (m:getKeywordParameters(m) | m <- models);
+  set[str] keys                  = {*domain(getKeywordParameters(m)) | m <- models };
```

Comment:
nit: this could be written as a comprehension over `fields` to avoid extra calls to `getKeywordParameters`.

Reply:
The current solution is faster, contra-intuitively. `getKeywordParameters` does not require a lookup in the hash trie, since this map is attached directly to the constructor. Especially for large sets of models this choice can save a few instructions per model.